engine/apps/phone_notifications/phone_backend.py

```diff
@@ -147,16 +162,109 @@ def notify_by_sms(self, user, alert_group, notification_policy):
 
         from apps.base.models import UserNotificationPolicyLogRecord
 
-        log_record_error_code = None
-
         renderer = AlertGroupSmsRenderer(alert_group)
         message = renderer.render()
+        _, log_record_error_code = self._send_sms(
+            user=user,
+            alert_group=alert_group,
+            notification_policy=notification_policy,
+            message=message,
+        )
 
+        if log_record_error_code is not None:
+            log_record = UserNotificationPolicyLogRecord(
+                author=user,
+                type=UserNotificationPolicyLogRecord.TYPE_PERSONAL_NOTIFICATION_FAILED,
+                notification_policy=notification_policy,
+                alert_group=alert_group,
+                notification_error_code=log_record_error_code,
+                notification_step=notification_policy.step if notification_policy else None,
+                notification_channel=notification_policy.notify_by if notification_policy else None,
+            )
+            log_record.save()
+            user_notification_action_triggered_signal.send(sender=PhoneBackend.notify_by_sms, log_record=log_record)
+
+    @staticmethod
+    def notify_by_sms_bundle_async(user, bundle_uuid):
+        notify_by_sms_bundle_async_task.apply_async((user.id, bundle_uuid))
+
+    def notify_by_sms_bundle(self, user, bundle_uuid):
+        """
+        notify_by_sms_bundle sends an sms notification bundle to a user using configured phone provider.
+        It handles business logic - limits, cloud notifications and UserNotificationPolicyLogRecord creation.
+        It creates UserNotificationPolicyLogRecord for every notification in bundle, but only one SMSRecord.
+        SMS itself is handled by phone provider.
+        """
+
+        from apps.alerts.models import BundledNotification
+        from apps.base.models import UserNotificationPolicy, UserNotificationPolicyLogRecord
+
+        notifications = BundledNotification.objects.filter(bundle_uuid=bundle_uuid).select_related("alert_group")
+
+        if not notifications:
+            logger.info("Notification bundle is empty, related alert groups might have been deleted")
+            return
+        channels_and_alert_groups_count = notifications.aggregate(
+            channels_count=Count("alert_receive_channel", distinct=True),
+            alert_groups_count=Count("alert_group", distinct=True),
+        )
+        # get 3 unique alert groups from notifications
+        alert_groups_to_render = []
+        for notification in notifications:
+            if notification.alert_group not in alert_groups_to_render:
```

Comment:
would a `set` work for this?

Reply:
yes, makes sense to use it instead 👍

Reply:
it this case it doesn't work, since it's not subscriptable